### PR TITLE
Implement Azure service connector

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
Implement the Azure service connector.

```
$ zenml service-connector list-types --type azure
┏━━━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━┯━━━━━━━┯━━━━━━━━┓
┃          NAME           │ TYPE     │ RESOURCE TYPES        │ AUTH METHODS      │ LOCAL │ REMOTE ┃
┠─────────────────────────┼──────────┼───────────────────────┼───────────────────┼───────┼────────┨
┃ Azure Service Connector │ 🇦 azure │ 🇦 azure-generic      │ implicit          │ ✅    │ ✅     ┃
┃                         │          │ 📦 blob-container     │ service-principal │       │        ┃
┃                         │          │ 🌀 kubernetes-cluster │ access-token      │       │        ┃
┃                         │          │ 🐳 docker-registry    │                   │       │        ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━┷━━━━━━━┷━━━━━━━━┛

$ zenml service-connector list
┏━━━━━━━━┯━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━┯━━━━━━━━┯━━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━┓
┃ ACTIVE │ NAME        │ ID                                   │ TYPE     │ RESOURCE TYPES        │ RESOURCE NAME │ SHARED │ OWNER   │ EXPIRES IN │ LABELS ┃
┠────────┼─────────────┼──────────────────────────────────────┼──────────┼───────────────────────┼───────────────┼────────┼─────────┼────────────┼────────┨
┃   👉   │ azure-multi │ 057ba070-c3d7-436b-87bc-f6334c35b240 │ 🇦 azure │ 🇦 azure-generic      │ <multiple>    │ ➖     │ default │            │        ┃
┃        │             │                                      │          │ 📦 blob-container     │               │        │         │            │        ┃
┃        │             │                                      │          │ 🌀 kubernetes-cluster │               │        │         │            │        ┃
┃        │             │                                      │          │ 🐳 docker-registry    │               │        │         │            │        ┃
┗━━━━━━━━┷━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━┷━━━━━━━━┷━━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━┛

$ zenml service-connector verify azure-multi
⠸ Verifying service connector 'azure-multi'...
Service connector 'azure-multi' is correctly configured with valid credentials and has access to the following resources:
┏━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃     RESOURCE TYPE     │ RESOURCE NAMES                                ┃
┠───────────────────────┼───────────────────────────────────────────────┨
┃   🇦 azure-generic    │ ZenML Subscription                            ┃
┠───────────────────────┼───────────────────────────────────────────────┨
┃   📦 blob-container   │ az://demo-zenmlartifactstore                  ┃
┠───────────────────────┼───────────────────────────────────────────────┨
┃ 🌀 kubernetes-cluster │ demo-zenml-demos/demo-zenml-terraform-cluster ┃
┠───────────────────────┼───────────────────────────────────────────────┨
┃  🐳 docker-registry   │ demozenmlcontainerregistry.azurecr.io         ┃
┗━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by AlexejPenner — APPROVED** [review]
Looks pretty good to me, that 1700 line file took some time to get through. Not sure I understood every smallest aspect of it so might make sense to also get a review from someone who reviewd the original connectors PR

File: src/zenml/integrations/kubernetes/service_connectors/kubernetes_service_connector.py
Comment (fa9r) [inline src/zenml/integrations/kubernetes/service_connectors/kubernetes_service_connector.py:265]: I see this logic duplicated several times below, should we move this to a separate util function perhaps? TBH I also don't fully understand what it does so having this in one place with corresponding docstrings would be helpful as well.

File: src/zenml/integrations/azure/service_connectors/azure_service_connector.py
Comment (fa9r) [inline src/zenml/integrations/azure/service_connectors/azure_service_connector.py:253]:
```suggestion
permissions of the managed identity may need to be adjusted to allow listing
```

File: src/zenml/integrations/azure/service_connectors/azure_service_connector.py
Comment (fa9r) [inline src/zenml/integrations/azure/service_connectors/azure_service_connector.py:319]:
```suggestion
are not represented by other, more specific resource types, like Azure blob
```

File: src/zenml/integrations/azure/service_connectors/azure_service_connector.py
Comment (fa9r) [inline src/zenml/integrations/azure/service_connectors/azure_service_connector.py:709]: include in docstrings above?

**Review by fa9r — COMMENTED** [review]
Overall this looks good to me, However, similar to @AlexejPenner, I found it quite hard to review this in detail without having reviewed the original connectors PR, so I would suggest waiting for a review from @schustmi (in addition to docs and tests) before merging this. 

High-level I was wondering why Azure/AWS/GCP all require k8s as a dependency now. In a sense this is not ideal since I would now need to install k8s even if I just want to use, e.g., the GCP artifact store. Do you think it might make sense to decouple the connectors from the existing integrations for that reason? Or perhaps to move the GCP/Azure artifact store implementations to separate integrations similar to how we have `aws` and `s3`?

Also, I think long texts like the `description` of your new models would be much easier to consume, review, and maintain if they lived in the docs. Thus, I'd propose to turn them into docs and to replace them with short one-line descriptions that include links to the relevant pages instead.

**Comment by stefannica** [comment]
> High-level I was wondering why Azure/AWS/GCP all require k8s as a dependency now. In a sense this is not ideal since I would now need to install k8s even if I just want to use, e.g., the GCP artifact store. Do you think it might make sense to decouple the connectors from the existing integrations for that reason? Or perhaps to move the GCP/Azure artifact store implementations to separate integrations similar to how we have `aws` and `s3`?

@fa9r  The kubernetes library is used in so many of our stack components that it's almost as fundamental to ZenML as the docker library. But more to the point, all these integrations are cloud provider integrations that include some form of managed Kubernetes and the connectors rely on the k8s library to generate kubernetes specific API credentials.

Decoupling connectors from integrations make it even more complicated to deal with dependencies, because the pipeline code now also needs to be aware of which connectors are linked to which components in order to track and install the necessary packages (just like we do with stack components and integrations). The compromise of including kubernetes as a dependency for all integrations that deal with cloud providers is the simplest way out of a situation that only exacerbates the dependency hell.

> 
> Also, I think long texts like the `description` of your new models would be much easier to consume, review, and maintain if they lived in the docs. Thus, I'd propose to turn them into docs and to replace them with short one-line descriptions that include links to the relevant pages instead.

I disagree. In fact, I think in a future where connectors and integrations exist separately from the ZenML core repository (i.e. Hub plugins), some form of documentation needs to live closer to the descriptors that encode what each plugin contains and how to use it. But more specifically for service connectors, the documentation embedded in the code is extensively used to provide a more interactive CLI experience and soon in more interactive dashboard experience as well.

You do have a point however that we do need to cover these service connectors in the ZenML docs and right now the real problem is that these same docs are duplicated there as well. A situation for which I don't have a solution at the moment, but maybe we can think of one as part of the future "zenml integrations as Hub plugins" movement.

**Review by fa9r — APPROVED** [review]
Looks good to me, I also don't have strong opinions about the general suggestions above, so happy to keep those as they are if you prefer that. 

Duplication between docs and src is a general problem we have so I don't expect that to be solved as part of this PR. Ideally, we should find a way to "import" text/code in our docs from other places but I think gitbook doesn't allow that, so nothing we can do about it right now I would say.

Thus, happy to approve, but, as I mentioned above, please wait for @schustmi's review before merging :)

File: src/zenml/integrations/azure/service_connectors/azure_service_connector.py
Comment (schustmi) [inline src/zenml/integrations/azure/service_connectors/azure_service_connector.py:1166]: I've seen this `split("/")[4]` in a few places now, should we extract it in some function in case it needs to be updated in the future?

**Review by schustmi — CHANGES_REQUESTED** [review]
Love it, had a quite detailed look but only have some minor comments! 🎉

(Apart from unit tests, but I completely understand that mocking the entire Azure framework would be a huge effort)

File: src/zenml/container_registries/azure_container_registry.py
Comment (stefannica) [reply src/zenml/container_registries/azure_container_registry.py:50]: I did it for the `docker-registry` part, but I can't import here constants from the Azure integration. Not sure how to solve this without polluting the global constants namespace with integration specific constants, which I'd rather not do.

File: src/zenml/integrations/azure/service_connectors/azure_service_connector.py
Comment (stefannica) [reply src/zenml/integrations/azure/service_connectors/azure_service_connector.py:673]: Yes, there are provisions in the ServiceConnector base class that validate the configuration before this function is even called: https://github.com/zenml-io/zenml/blob/main/src/zenml/models/service_connector_models.py#L193

File: src/zenml/container_registries/azure_container_registry.py
Comment (schustmi) [reply src/zenml/container_registries/azure_container_registry.py:51]: Ah I see, because this file is still in core ZenML, makes sense

**Review by schustmi — APPROVED** [review]

**Pull request merged**
